proxy_connect.go

```diff
@@ -0,0 +1,170 @@
+package proxy
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/getlantern/errors"
+	"github.com/getlantern/hidden"
+	"github.com/getlantern/idletiming"
+	"github.com/getlantern/netx"
+)
+
+// BufferSource is a source for buffers used in reading/writing.
+type BufferSource interface {
+	Get() []byte
+	Put(buf []byte)
+}
+
+// CONNECT returns a Handler that processes CONNECT requests.
+//
+// idleTimeout: if specified, lets us know to include an appropriate
+// KeepAlive: timeout header in the CONNECT response
+//
+// bufferSource: specifies a BufferSource, leave nil to use default
+//
+// dial: the function that's used to dial upstream
+func CONNECT(
+	idleTimeout time.Duration,
+	bufferSource BufferSource,
+	dial DialFunc,
+) Interceptor {
+	// Apply defaults
+	if bufferSource == nil {
+		bufferSource = &defaultBufferSource{}
+	}
+
+	ic := &connectInterceptor{
+		idleTimeout:  idleTimeout,
+		bufferSource: bufferSource,
+		dial:         dial,
+	}
+	return ic.connect
+}
+
+// interceptor configures an Interceptor.
+type connectInterceptor struct {
+	idleTimeout  time.Duration
+	bufferSource BufferSource
+	dial         DialFunc
+}
+
+func (ic *connectInterceptor) connect(w http.ResponseWriter, req *http.Request) error {
+	var downstream net.Conn
+	var upstream net.Conn
+	var err error
+
+	closeDownstream := false
+	closeUpstream := false
+	defer func() {
+		if closeDownstream {
+			if closeErr := downstream.Close(); closeErr != nil {
+				log.Tracef("Error closing downstream connection: %s", closeErr)
+			}
+		}
+		if closeUpstream {
+			if closeErr := upstream.Close(); closeErr != nil {
+				log.Tracef("Error closing upstream connection: %s", closeErr)
+			}
+		}
+	}()
+
+	// Hijack underlying connection.
+	downstream, _, err = w.(http.Hijacker).Hijack()
+	if err != nil {
+		fullErr := errors.New("Unable to hijack connection: %s", err)
+		respondBadGateway(w, fullErr)
+		return fullErr
+	}
+	closeDownstream = true
+
+	// Note - for CONNECT requests, we use the Host from the request URL, not the
+	// Host header. See discussion here:
+	// https://ask.wireshark.org/questions/22988/http-host-header-with-and-without-port-number
+	upstream, err = ic.dial("tcp", req.URL.Host)
+	if err != nil {
+		fullErr := errors.New("Unable to dial upstream: %s", err)
+		ic.respondBadGatewayHijacked(downstream, req, fullErr)
```

Comment:
Maybe not worth doing it, but if we swap the order of hijacking client connection and dialing origin, we can use `respondBadGateway` here and remove `respondBadGatewayHijacked` altogether. `Hijack` operation rarely fail, so it won't waste time to dial origin in most cases.